design/open/branch_checks.md

```diff
@@ -0,0 +1,171 @@
+```
+
+#### Check event syntax
+
+```json
+{
+  "repository_id": "example-repo",
+  "branch_id": "branch_a",
+  "source_ref": "0307e79ab5afef2b0726a8c13896f716033b1acd8454e22210bd43a2f72a94a9",
+  "check_id": "great_expectations_validate_events",
+  "storage_namespace": "s3://path/to/storage/",
+  "callback_token": "abc123",
+  "output_url": "presigned URL or another HTTP POST-able URL"
+}
+```
+
+#### Check triggering
+
+Checks are triggered when the `check(repository_id, ref_id)` endpoint is called in the lakeFS API.
+When running check() with a branch or tag, it will be dereferenced: checks are valid for a given commit ID.
+
+
+#### Check execution
+
+Like hooks, checks could have multiple type implementations (Airflow, Webhooks, etc). Initially, only Airflow is supported.
+
+`execute()` will trigger a new [DAG run using the Airflow REST API](https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/get_dag_runs), passing as conf the settings as shown above (including the callback token)
+
+#### Check status
+
+```mermaid
+stateDiagram-v2
+    direction LR
+    [*] --> STARTING: execute()
+    STARTING --> FAILED: error calling Airflow
+    STARTING --> EXECUTING
+    EXECUTING --> SUCCESS: callback(success)
+    EXECUTING --> FAILED: callback(error)
+    EXECUTING --> LOST: timeout
+    LOST --> STARTING: retry
+    FAILED --> STARTING: retry
+    SUCCESS --> [*]
+```
+
+#### Retrying checks
+
+From the `lost` and `failed` statuses, users can trigger a retry operationthat will cause the execute function to run again for the given check.
+This is used when transient errors occur (or a timeout).
+
+#### Getting check status
+
+##### `/api/v1/repositories/{repoID}/refs/{refID}/checks`
+
+```json
+{
+  "checks": [
+    {"id": "great_expectations_validate_events", "status": "SUCCESS", "execution_id": "1234"},
+    {"id": "other_check", "status": "EXECUTING", "execution_id": "45678"},
+    {"id": "yet_another_check", "status": "LOST", "execution_id": "8900"},
+  ]
+}
+```
+
+#### Callback API
+
+Each execution will send a `callback_token` to the executing system (Airflow, for now).
+
+`POST /api/v1/repositories/{repoID}/refs/{refID}/checks/{checkID}?token={callback_token}`
+
+```json
+{
+  "status": "SUCCESS",
+  "metadata": {},
+}
+```
+
+Where metadata is a simple `Map<string, string>` mapping to be sent by the executing system.
+
+When lakeFS receives a callback, it checks that the provided token is the last one issued for that (check, commit). 
+Any previous token issued will be ignored.
+
+#### Merging into a protected branch
+
+A branch may define 0 or more checks that must be in a state of SUCCESS for the ref being merged into it.
+Once a merge is issued, the server will dereference the source branch and get back its list of checks to validate this status.
+
+This could be done using a pre-merge hook that runs this validation, or at a higher level using the branch protection settings.
+
```

Comment:
1. Checks don't run unless you trigger them.
2. When running checks, you don't yet know which branch you're merging to. Mandatory checks are defined on the receiver to separate concerns. Let's take a real world example:

Let's say I have a `users` table that has an `age` column. I can define a check on `main` whose ID is `ensure_age_is_between_0_and_125`. I can check this with a Spark job, with Great Expectations or with DuckDB, that's an implementation detail. It's up to the source branch to make sure this check passes and how it is implemented. 

This is also the case for e.g. GitHub actions - defining the mandatory check is a setting on the receiving branch, not within the action file itself.

Reply:
3. For GitHub PR's that's not the case. If you create a PR, all defined actions that trigger before a merge would happen. If you push another commit to the source branch of the PR - your actions would run again. 

Branch checks in lakeFS have the potential of being much more expensive - running a distributed system and potentially scanning large datasets. I don't think this behavior makes much sense there - this is why check() is an independent API and doesn't trigger automatically.

Reply:
Thanks for clarifying, I still don't get on how the destination branch enforces the check on the source branch if they may diverge and apply different logic.
Example:
1. `main`: check `ensure_age`: validates the age is between 0 and 125.
2. created a branch `feature_branch`.
3. `main`: committed a change to the check that the age is now between 0 and 100.
4. `feature_branch`: committed insertion of ages between 100-125 and triggered `check()`. It passed.
5. Merge from `feature_branch` to `main` is successful since the required check passed.